### PR TITLE
Added BigQuery support

### PR DESCRIPTION
**Pull request by azhard**
## Description & motivation
Added support for the event-logging package with BigQuery dbt projects.

Key changes:
- Updated dbt and dbt_utils version requirements to leverage `adapter.dispatch` functionality
- Added `bigquery.sql` file that overwrites some of the default macros
  - Key differences being no `event_user` column and no use of `commit`
- Added `bigquery` models folder to account for schema differences
- Updated `create_audit_schema` macro to use the underlying dbt `create_schema` macro instead of creating the schema using raw SQL
- Added `log_custom_event` macro in case a user wants to log their own event types
- Added CircleCI integration test for BigQuery (although don't think I can trigger these to verify they work)

Note: local testing worked great but only when threads was set to 1. I added this callout in the `README` file.

## Checklist
- [X] I have verified that these changes work locally
- [X] I have updated the README.md (if applicable)
- [X] I have added tests & descriptions to my models (and macros if applicable)

### CODE REVIEW

**Comment by azhard** [comment]
It would be possible to add the changes that bring support to BigQuery without the need to upgrade to dbt 0.18.0 but I figured at some point the minimum version would have to be bumped regardless. However if that's something that reviewers think should be done at a later point in time, I can refactor the PR to have it support pre 0.18.0 releases

File: integration_tests/macros/create_old_audit_table.sql
Comment (jtcohen6) [inline integration_tests/macros/create_old_audit_table.sql]: Since `dbt_utils` is already a dependency for this package, I figure all of these should be
```suggestion
       event_name       {{ dbt_utils.type_string() }},
```
and then you can skip the logic above

**Review by jtcohen6 — COMMENTED** [review]
One tiny comment, and I kicked off integration tests (which aren't currently set up to run from forks).

@azhard If there's anyone I'd want to stress-test this package on BQ, it's you. We've found that, while the resulting information is useful in principle, the sort of micro-batch insertions this package asks the warehouse to do are _not_ performant at scale. In many projects, the loss of efficiency ends up outweighing the benefits. That's the big reason why, frankly, we've hesitated to invest meaningful resources into its upkeep or recommend it widely for its stated purpose.

There are two other places to look today for information on dbt runtime performance:
- **Warehouse query logs** (in BQ's case, [monitoring](https://cloud.google.com/bigquery/docs/monitoring) + `labels` for ease of parsing/filtering). I think there's a lot of potential here, and there's significant overlap with broader efforts to understand _all_ the downstream uses of dbt resources (related thread: https://github.com/fishtown-analytics/dbt/issues/2730#issuecomment-693393683).
- **dbt-produced artifacts**, namely `run_results.json`.

You can see an initial attempt at the latter (using the information available in the end-of-run Jinja context) in #12. In order for that approach to be plausible, we need:
- a stable schema for those artifacts (coming in v0.19.0)
- a package/playbook of recommended transformations on top of those artifacts, e.g. to identify runtime variance, development cadence, column structure variability, ...
- (if using dbt Cloud) a mechanism for ingesting them back into the warehouse

Those are things we _want_ to have, someday, as a better answer to the problem this package is trying to solve. Until then, I'm happy to merge a good PR :)

File: integration_tests/macros/create_old_audit_table.sql
Comment (azhard) [reply integration_tests/macros/create_old_audit_table.sql]: I was thinking of that as well but wasn't sure if it would break anything on the non-BigQuery side as `dbt_utils.type_string()` would evaluate to `varchar` instead of `varchar(512)` for other DBs.

Just want to confirm that you still think that's an okay change to make?

**Comment by azhard** [comment]
As far as testing goes, the runtime has definitely increased especially because implementing this would cause us to change threads from 4 to 1. As for an intense stress test, I'm not sure I can truly vouch for that performance as our dbt project only has about 40 models.

The rest of what you said makes a lot of sense to me and I'm really excited to see the evolution of `reports`. I think leveraging the results directly from the data warehouse makes a ton of sense intuitively.

My issue with the `run_results.json` method is that our team is hoping to use the dbt logged events to not only identify runtimes and the like, but to also eventually build a dashboard that shows us:
1.  how long a model takes to deploy on average
2. when was the last time it was deployed
3. what models are currently being deployed

I'm sure that list will grow longer in the future but `3` provides the real limitation with the run_results method as the results are only written at the end of the run vs. after each successful model has been run.

Always happy to contribute as well, so if there's anyway I'm able to help implement any of the proposed changes feel free to let me know!

Also, pushed an update that should 🤞 fix the integration tests.

File: dbt_project.yml
Comment (jtcohen6) [inline dbt_project.yml]: I think you should `as_bool` these to be safe:
```suggestion
      +enabled: '{{ target.type == "bigquery" | as_bool }}'
    default:
      +enabled: '{{ target.type != "bigquery" | as_bool }}'
```

File: integration_tests/macros/create_old_audit_table.sql
Comment (jtcohen6) [reply integration_tests/macros/create_old_audit_table.sql]: Yes! Redshift, Postgres, and Snowflake are all okay with setting column types to `varchar` with no length specification. And if they did need a specification, that should be a matter for the `dbt_utils.type_string()` definition IMO

File: macros/audit.sql
Comment (jtcohen6) [inline macros/audit.sql]:
```suggestion
        database=target.database,
```
This is what's causing the integration test error currently. `target.project` aliases to `target.database` for the `dbt-bigquery` plugin only.

**Review by jtcohen6 — COMMENTED** [review]
Sorry for the delay @azhard! I found one other spot that's tripping up the integration tests.

Also, thanks for the context on what you're after. This is the conversation I'm _really_ interested in :)

> 3. what models are currently being deployed

Really fair, `run_results.json` cannot get you this kind of information intra-run. I do see this as a different category of problem (operational) vs. 1 (longitudinal), with 2 somewhere in between.

I wonder if query logs could be the angle here instead. What do you think about having your dashboard "pull" from `information_schema.jobs_by_*` on an interval (whether 30s or 5m), filtering on a user or query label as appropriate, and parse the model name out of the [query comment `node_id`](https://docs.getdbt.com/reference/project-configs/query-comment/#default)?

**Comment by azhard** [comment]
This is actually my first time learning about `information_schema.jobs_by_*` in BigQuery. I actually feel that using this and building models on top of that can solve almost all of the use cases I had in mind.

I think it would be really interesting if `dbt` had suggested models built on top of that "table" but that's obviously more a future enhancement.

For now, I think on my side I'm okay to put this addition to the `event-logging` repo on hold. If you still think it's worth merging in case anyone else wants to use it in the future, I've updated the MR based on your comments.

**Comment by azhard** [comment]
I'd also add that it would be great to have the dbt invocation_id included in the query-comment. Happy to create an MR to do that if you think it'd be worth adding as well

**Comment by jtcohen6** [comment]
> I think it would be really interesting if `dbt` had suggested models built on top of that "table" but that's obviously more a future enhancement.

I really, really want to do this. Perhaps as a package? (Perhaps as a total rework of _this_ package?) I think database compatibility makes this trickier, but also all the more worth doing.

If you end up writing queries/models you're happy with, I'd love to give them a home.

> I'd also add that it would be great to have the dbt invocation_id included in the query-comment. Happy to create an MR to do that if you think it'd be worth adding as well

- ~The `invocation_id` _is_ included in the query-comment!~
- As of v0.19.0, the `invocation_id` will also be included as a job label for all dbt-executed queries (#2809)

**Comment by azhard** [comment]
Was the `invocation_id` added to the query-comment in 0.19.0 as well? Current comment I'm seeing is along the lines of 
```
/* {"app": "dbt", "dbt_version": "0.18.0", "profile_name": "dbt", "target_name": "dev", "node_id": "model.dbt.test_model"} */
```
And the [default in the docs](https://docs.getdbt.com/reference/project-configs/query-comment/#default) doesn't show the `invocation_id` either

**Comment by jtcohen6** [comment]
You're right, I'm not sure why I imagined it to be in the default. In any case, you can define a `query_comment` macro to override the default, and its available [compilation context](https://docs.getdbt.com/reference/project-configs/query-comment/#compilation-context) includes `invocation_id`.

File: macros/audit.sql
Comment (jtcohen6) [inline macros/audit.sql]:
```sql
       ["event_name", dbt_utils.type_string()],
       ["event_timestamp", dbt_utils.type_timestamp()],
       ["event_schema", dbt_utils.type_string()],
       ["event_model", dbt_utils.type_string()],
       ["event_user", dbt_utils.type_string()],
       ["event_target", dbt_utils.type_string()],
       ["event_is_full_refresh", "boolean"],
       ["invocation_id", dbt_utils.type_string()],
```
sorry if I led you astray with my previously suggested code here

Comment (azhard) [reply macros/audit.sql]: Good catch, updated the PR

**Review by jtcohen6 — COMMENTED** [review]
Super close. I think you just need to add a `setup_creds` step to `circleci/config.yml`, between `checkout` and `restore_cache`:
```
      - run:
          run: setup_creds
          command: |
            echo $BIGQUERY_SERVICE_ACCOUNT_JSON > ${HOME}/bigquery-service-key.json
```

(the same code as [here](https://github.com/fishtown-analytics/snowplow/blob/f24a2bf91d4ce44f789d1cae0e33d85aa7f8eb58/.circleci/config.yml#L13-L16))

**Comment by azhard** [comment]
> Super close. I think you just need to add a `setup_creds` step to `circleci/config.yml`, between `checkout` and `restore_cache`:
> 
> ```
>       - run:
>           run: setup_creds
>           command: |
>             echo $BIGQUERY_SERVICE_ACCOUNT_JSON > ${HOME}/bigquery-service-key.json
> ```
> 
> (the same code as [here](https://github.com/fishtown-analytics/snowplow/blob/f24a2bf91d4ce44f789d1cae0e33d85aa7f8eb58/.circleci/config.yml#L13-L16))

Updated

File: .circleci/config.yml
Comment (jtcohen6) [inline .circleci/config.yml]: hope this is the final piece
```suggestion
         name: "Run Tests - BigQuery"
         environment:
              BIGQUERY_SERVICE_KEY_PATH: "/home/circleci/bigquery-service-key.json"
```
[a la](https://github.com/fishtown-analytics/snowplow/blob/master/.circleci/config.yml)

**Review by jtcohen6 — APPROVED** [review]
Thanks for all the hard work here @azhard !!!

**Pull request merged**
